### ISSUE TITLE
Revoking a grain sharing link revokes access for users who used that link

### ISSUE THREAD

**Issue by nathanhourt** [sandstorm-io/sandstorm#826]
If Alice creates a grain G, and shares it with Bob via a sharing link, then revokes that sharing link, Bob's access to G is revoked. This is undesirable; if Alice wanted to revoke Bob's access, she should do that as a separate step (perhaps prompted when she revokes the link, but it should also be available later).

This is closely related to #701 which describes the necessity of using a side-channel to share grains. Since that side-channel may not be secure, an unauthorized eavesdropper Eve could gain access to G by using the link. While Eve's unauthorized access could be discovered, at that point the damage would already be done, and it should not be a possibility to begin with.

In order to minimize this risk, it would be desirable that the link could be revoked as soon as the authorized user used it to gain access to the grain, thus minimizing the window for attack; however, this does not appear to be possible as revoking the link also revokes the authorized user's access.

**Comment by ocdtrekkie** [comment]
So, without knowing if it's possible to even do this: Perhaps if the share by email tab is used, it should invalidate the link after it's used once, because the user emailed is presumably the one who got it. But if you 'create a sharing link', that sticks around no matter how many people use it, because presumably you wanted to post.

**Comment by kentonv** [comment]
Yes, we plan to extend the sharing UI to support revoking a link without revoking the people who received access through it. It's tricky, though, because there is also a strong use case for transitive revocation: say you accidentally leaked a link and people have already gotten access through it. You now want to revoke all those people without revoking people who legitimately had access without the link. So we will need to prompt for this.

This change will probably come together with share-by-identity, which will make it possible to share directly to another Sandstorm user without a link in between. Probably the implementation will be that when you revoke the link but request to keep the existing users, we'll actually change the sharing graph as if you had shared directly to all those users by identity.

@ocdtrekkie single-use links don't work because people often address share mails to mailing lists. (This is something I learned while working on Google Docs sharing.)

**Comment by dwrensha** [comment]
Unassigning myself because I am not planning to work on this within the next two weeks.
